### ISSUE TITLE
The contact request is not being resent if previous was declined

### ISSUE THREAD

**Issue by pavloburykh** [status-im/status-react#14798]
**Steps:**
- User A: send CR
- User B: decline it
- User A: send CR again
**Expected result:**
CR will be received
**Actual result:**
no CR received

Issues is likely to be a consequence of https://github.com/status-im/status-mobile/pull/14747. It is a blocker for https://ethstatus.testrail.net/index.php?/tests/view/3075386 

### Additional Information

- Status version: nightly 
- Operating System: Android, iOS

[comment]: # (Please, add logs/notes if necessary)

**Comment by cammellos** [comment]
@pavloburykh this might be expected behavior, if you decline a contact request, you won't receive a notification anymore if the other contact sent it.
 
@ilmotta @John-44 @xAlisher what's the expected behavior?

**Comment by pavloburykh** [comment]
> @pavloburykh this might be expected behavior, if you decline a contact request, you won't receive a notification anymore if the other contact sent it.

@cammellos current implementation does not let users to become mutual contacts in some easy intuitive way in case if CR has been once declined.

Here is a case:

1. User A sends CR to User B
2. User B declines CR - **as a result notifications will not appear for User B if User A re-sends CR**
3. User B sends CR to User A - **as a result, User A does not receive any CR notifications but User B automatically appears in contact list of User A. User B in his turn is unable to send messages to User A, message input is blocked as CR is pending. User A does not appear in contact list for User B.** 
4. Users will become mutual contacts only after User A will send CR once again to User B. 

Such behaviour does not seem to be correct. 

Also, there might be situations when User declines CR by a mistake and there must be some easy intuitive way for users to become mutual contacts.  


https://user-images.githubusercontent.com/97245802/215069242-38e849d6-08f6-48ac-b069-f8ffc0921057.mp4

**Comment by ilmotta** [comment]
@cammellos, I've always seen this behavior as an old bug that I didn't bother bringing up. Thanks @pavloburykh. It's been like this ever since I tested CR flows for the first time.

I agree with @pavloburykh, it's going to be frustrating for users if they can't be mutual contacts if the receiver declines the CR. In my local experiments I always ended up creating new accounts to overcome this limitation, but that won't be an option for real users.

It's like if user A declines a CR from B, it behaves as "block user B forever and I don't want to hear from that person ever again" ;) I understand the reasoning for avoiding spamming users with contact requests, so I'm not really sure how we should go about this, there's more than one solution. I guess we should take this problem to designers.

**Comment by cammellos** [comment]
> @cammellos, I've always seen this behavior as an old bug that I didn't bother bringing up. Thanks @pavloburykh. It's been like this ever since I tested CR flows for the first time.
> 
> I agree with @pavloburykh, it's going to be frustrating for users if they can't be mutual contacts if the receiver declines the CR. In my local experiments I always ended up creating new accounts to overcome this limitation, but that won't be an option for real users.
> 
> It's like if user A declines a CR from B, it behaves as "block user B forever and I don't want to hear from that person ever again" ;) I understand the reasoning for avoiding spamming users with contact requests, so I'm not really sure how we should go about this, there's more than one solution. I guess we should take this problem to designers.

I am ambivalent, as long as the solution makes sense it's all good with me, either dismiss acts similarly to "block', but then there's a way to recover as @pavloburykh correctly pointed out, or dismiss just removes the notification, and it shows again on the next contact request, it's up to us. 

I think 1 is preferable as we don't allow contacts to resend requests if they have been dismissed, as they are not notified, for privacy/spam reasons, so having a section where you could see the dismissed contacts and revert the action seems more inline, and it's similar to how facebook for example handles contacts/friends request.

The question is also, what we do for RC, just to make sure it's usable (whether this would be considered a blocker, I am not sure to be honest, but worth discussing)

**Comment by ilmotta** [comment]
@cammellos, @pavloburykh: thinking in more detail now, if user `A` declined a CR from `B`, and if they really want to establish mutual contacts, wouldn't it just work now if `A` sent a CR to `B` (the reverse)? If this works then I think it's a reasonable middle ground for the RC. But if both of them decline the CR (`A->B` and `B->A`), then they're out of luck.

I think the ideal solution should somehow empower the user to decide if they want to re-accept CRs from anybody. I can imagine a feature where a user would be able to explicitly add a public key from someone they once declined a CR.

**Comment by cammellos** [comment]
> If this works then I think it's a reasonable middle ground for the RC. But if both of them decline the CR (A->B and B->A), then they're out of luck.

That's how it works, correct, it should make the two contacts mutual.

**Comment by pavloburykh** [comment]
>  If this works then I think it's a reasonable middle ground for the RC. 

@ilmotta I agree, that would work as for RC 1. And in future we can improve the flow, basing on designers solution which will be provided for this case. 

But current behaviour does not correspond to what we expect and should be improved. As I have mentioned [above](https://github.com/status-im/status-mobile/issues/14798#issuecomment-1406335278):

> 3. User B sends CR to User A - **as a result, User A does not receive any CR notifications but User B automatically appears in contact list of User A. User B in his turn is unable to send messages to User A, message input is blocked as CR is pending. User A does not appear in contact list for User B.

You can observe this behaviour on video which I have attached to comment above.

**Comment by cammellos** [comment]
@pavloburykh oh, that definitely needs fixing, good catch!

Could you check if the same behavior happens on https://github.com/status-im/status-mobile/pull/14878 please?

Thank you

**Comment by pavloburykh** [comment]
> Could you check if the same behavior happens on #14878 please?

Sure, I will check!

**Comment by John-44** [comment]
@cammellos @pavloburykh @ilmotta So this is what I think should happen:

1) Alice sends a contact request to Bob

2) Bob declines the contact request 

3) Alice sees that the contact request she sent to Bob is still (from her point of view) in the pending state

4) Alice should not be able to send another contact request to Bob (because in her UI her original contract request to Bob is still pending). If she could send more contact requests to Bob, this would be a spam vector.

5) At any point in the future, Bob (if he wishes to change his mind) can go into Settings/Messaging/Contacts where he can change his earlier Rejection into an Acceptance of Alice’s earlier contact request. Assuming Bob does this, then:

6) Alice will see the contact request she previously sent to Bob change to the Accepted state, and Alice and Bob will now be mutual contacts

**Comment by pavloburykh** [comment]
> @cammellos @pavloburykh @ilmotta So this is what I think should happen:

Yeah, such flow looks correct to me. But do we want to implement such flow for RC-1 or this will be done in future releases and for now we are good with [flow](https://github.com/status-im/status-mobile/issues/14798#issuecomment-1406384153) described by @ilmotta ?

**Comment by cammellos** [comment]
> > @cammellos @pavloburykh @ilmotta So this is what I think should happen:
> 
> Yeah, such flow looks correct to me. But do we want to implement such flow for RC-1 or this will be done in future releases and for now we are good with [flow](https://github.com/status-im/status-mobile/issues/14798#issuecomment-1406384153) described by @ilmotta ?

Both apply I believe, on RC, the only way to recover will be the flow described by @ilmotta , while eventually we'll add the screen that @John-44 mentioned, but won't make the cut for RC.

**Comment by pavloburykh** [comment]
> Could you check if the same behavior happens on #14878 please?

The behaviour almost corresponds to what we expect except one small issue which we might need to improve: 

### Unwanted CR notification appears for User A, after User B (who initially declined User's A CR) sends CR to User A 

**Steps:**

1. User A send CR to User B
2. User B declines User's A CR
3. User B sends CR to User A

**Actual result:** Users A and B become mutual contacts, but unwanted CR notification appears for User A.

**Expected result:** we don't need to show CR notification for User A, as in this case Users already become mutual contacts and there is not need to accept it by User A. 


https://user-images.githubusercontent.com/97245802/215090962-9bcd1fb3-7e87-4e13-bcbd-2c9dcab13dec.mp4


@cammellos I can log it in https://github.com/status-im/status-mobile/pull/14878 or we can handle it separately. WDYT?

**Comment by cammellos** [comment]
@churik I believe this has been logged already, if that's the only issue with the PR, Id' be fixing it separately, but I can't find the issue anymore :)

**Comment by pavloburykh** [comment]
> @churik I believe this has been logged already, if that's the only issue with the PR, Id' be fixing it separately, but I can't find the issue anymore :)

@cammellos sorry, I am still in process of testing the PR. @churik has tested most cases but there is part of things left that need to be tested. As she is off toady she asked me to finish testing. I will keep you updated on progress. 

Regarding this [issue](https://github.com/status-im/status-mobile/issues/14798#issuecomment-1406469426) I am not sure that it has been logged before. But I agree that it can be addressed separately.

**Comment by cammellos** [comment]
> @cammellos sorry, I am still in process of testing the PR.

Oh yes, of course, I wasn't implying that I wanted to merge now, apologies if it sounded that way, thanks for testing!

**Comment by pavloburykh** [comment]
> Oh yes, of course, I wasn't implying that I wanted to merge now, apologies if it sounded that way, thanks for testing!

oh, not at all, didn't sound like that) I just wanted to let you know that I am still working on that PR. Thank you!

**Comment by cammellos** [comment]
@pavloburykh could you recheck please https://github.com/status-im/status-mobile/issues/14798#issuecomment-1406335278 on develop? 

https://github.com/status-im/status-mobile/issues/14798#issuecomment-1406335278 should be fixed, so in theory, we should be good for RC, if so, would you mind closing this issue please?
Thank you!

**Comment by pavloburykh** [comment]
> @pavloburykh could you recheck please [#14798 (comment)](https://github.com/status-im/status-mobile/issues/14798#issuecomment-1406335278) on develop?

As new "Add contact" flow has been already merged into develop case with decline seems to be different from what we expect according to this comment https://github.com/status-im/status-mobile/issues/14798#issuecomment-1406384153

**What we have now:**

1. In case if User A declines CR of User B, User B is still able to spam User A with contact requests by removing and re-adding User A to contacts using profile page (see video below). 

2. In case if User A who initially declined User's B CR sends CR to User B -  users will become mutual contacts, as expected. What is still unexpected - is this Unneeded CR that will be received by User B. We have discussed it here https://github.com/status-im/status-mobile/issues/14798#issuecomment-1406469426 and agreed to fix it separately. We really might want to fix this issue for RC-1 as it causes additional unexpected behavior in case if User B declines User's A CR: User A will be removed from User's B contact list, but User B will remain as a contact for User A. 

In this video you can observe cases which are described above. 
Based on that we need to decide if current behaviour is acceptable for rc-1

https://user-images.githubusercontent.com/97245802/215843363-93fbcce2-40be-45c5-ab03-bd13b5557823.mp4

[USER_A.zip](https://github.com/status-im/status-mobile/files/10549440/USER_A.zip)
[USER_B.zip](https://github.com/status-im/status-mobile/files/10549441/USER_B.zip)

**Comment by cammellos** [comment]
> In case if User A declines CR of User B, User B is still able to spam User A with contact requests by removing and re-adding User A to contacts using profile page (see video below).

This needs fixing, it's probably ok for RC since it's not critical, but if user declines as per above it should not pop up another request. Thank you

> https://github.com/status-im/status-mobile/issues/14798#issuecomment-1406469426 and agreed to fix it separately.

Yes, that should make it for the RC cut for sure

thanks!

**Comment by pavloburykh** [comment]
@cammellos thanx for clearing out!

> This needs fixing, it's probably ok for RC since it's not critical, but if user declines as per above it should not pop up another request.

Logged a separate issue which can be addressed in some future releases https://github.com/status-im/status-mobile/issues/14947

> Yes, that should make it for the RC cut for sure

Logged a separate issue for rc-1 https://github.com/status-im/status-mobile/issues/14946

Closing this issue.

**Comment by John-44** [comment]
@cammellos @pavloburykh @ilmotta see my reply to @MishkaRogachev here https://github.com/status-im/status-desktop/issues/9505#issuecomment-1425769497

I think the above stems from a bit of a misunderstanding of the designs (which is the fault of the desktop designs because the desktop designs are not very clear), I hope my comment on this issue clears this misunderstanding up
